File: SudokuKata/SudokuKataTests/ProgramTests.cs
using ApprovalTests;
using SudokuKata;
using System;
using System.IO;
using ApprovalTests.Namers;
using SudokuKataTests.TestHelpers;
using Xunit;
using Xunit.Abstractions;

namespace SudokuKataTests
{
    [Collection("Sequential")]
    //  ^^^ Prevent running in parallel with other tests that write to console
    //  This test works by capturing any output to standard output.
    //  It must not run in parallel with any other tests that also write to standard output.
    //  Any other tests that also capture console output should also be put in the same Collection.
    //  See https://stackoverflow.com/a/47174842/104370
    public class ProgramTests
    {
        private readonly ITestOutputHelper _testOutputHelper;

        public ProgramTests(ITestOutputHelper testOutputHelper)
        {
            _testOutputHelper = testOutputHelper;
        }

        [Fact]
        public void TestSeed()
        {
            var errors = 0;
            for (int seed = 0; seed < 20; seed++)
            {
                var randomValueGenerator = new FileWritingRandomNumber(seed);

                var section = SeedSectionName(seed);
                using (var cleanup = NamerFactory.AsEnvironmentSpecificTest("" + section))
                {
                    try
                    {
                        VerifySudokuForSeed(seed, randomValueGenerator);
                    }
                    catch (Exception e)
                    {
                        _testOutputHelper.WriteLine(e.ToString());
                        errors += 1;
                    }
                }

                using (var cleanup = NamerFactory.AsEnvironmentSpecificTest("" + section))
                {
                    var seedsFile = GetSeedsFileName(seed);
                    randomValueGenerator.WriteValuesToFile(seedsFile);
                }
            }
            Assert.Equal(0, errors);
        }

        private static string SeedSectionName(int seed)
        {
            return $"{seed:D4}";
        }

        private static void VerifySudokuForSeed(int seed, FileWritingRandomNumber randomValueGenerator)
        {
            var currentConsoleOut = Console.Out;
            using (var consoleOutput = new ConsoleUtilities.ConsoleOutput())
            {
                Program.Play(randomValueGenerator);
                string s = consoleOutput.GetOuput();
                Approvals.Verify(s);
            }
            Assert.Equal(currentConsoleOut, Console.Out);
        }

        private static string GetSeedsFileName(int seed)
        {
            var namer = Approvals.GetDefaultNamer();
            var basename = Path.Combine(namer.SourcePath, namer.Name);
            var seedsFile = basename + ".seeds.txt";
            return seedsFile;
        }
    }
}

File: SudokuKata/SudokuKataTestsCpp/ProgramTests.cpp
#include "ProgramTests.h"

#include "catch2/catch.hpp"
#include "ApprovalTests.hpp"

#include "Program.h"
#include "Utilities/FileReadingRandomNumber.h"
#include "Utilities/WStringHelpers.h"
#include "TestHelpers/FirstNFailuresReporter.h"

#include <iostream>
#include <regex>

using namespace ApprovalTests;
using namespace SudokuKata;

namespace SudokuKataTests
{

    TEST_CASE("TestSeed")
    {
        auto firstN = std::make_shared<FirstNFailuresReporter>(5, new DiffReporter);
        auto defaultReporterDisposer = Approvals::useAsDefaultReporter(firstN);
        auto errors = 0;
        for (int seed = 0; seed < 20; seed++)
        {
            // var randomValueGenerator = new FileWritingRandomNumber(seed);
            auto randomValueGenerator = new FileReadingRandomNumber();

            std::string section = ProgramTests::SeedSectionName(seed);
            auto sectionDisposer = NamerFactory::appendToOutputFilename(section);

            {
                // std::cout << "=============================================================" << std::endl;
                randomValueGenerator->ReadValuesFromFile(ProgramTests::GetSeedsFileName());
            }

            {
                try
                {
                    //                    auto defaultReporterDisposer =
                    //                        Approvals::useAsDefaultReporter(std::make_shared<QuietReporter>());

                    ProgramTests::VerifySudokuForSeed(randomValueGenerator);
                }
                catch (const std::exception& e)
                {
                    // The Approval Test failed - log the output on the console, and continue.
                    // so that we can review all failures in one session:
                    std::cout << "ERROR: " << e.what() << std::endl << std::endl;
                    errors += 1;
                }
            }

            // {
            //     var seedsFile = GetSeedsFileName(seed);
            //     randomValueGenerator.WriteValuesToFile(seedsFile);
            // }
        }
        CHECK(0 == errors);
    }

    std::string ProgramTests::SeedSectionName(int seed)
    {
        std::stringstream s;
        s << std::setfill('0') << std::setw(4) << seed;
        return s.str();
    }

    void ProgramTests::VerifySudokuForSeed(SudokuKata::IRandomValueGenerator* randomValueGenerator)
    {
        std::stringstream console;
        try
        {
            Program::Play(console, randomValueGenerator);
            std::string s = console.str();
        }
        catch (const std::exception& e)
        {
            console << "ERROR: " << e.what() << std::endl;
        }
        Approvals::verify(console.str());
    }

    std::string ProgramTests::GetSeedsFileName()
    {
        auto namer = Approvals::getDefaultNamer();
        auto seedsFile = toWString(namer->getApprovedFile(".seeds.txt"));
        return std::regex_replace(seedsFile, std::regex("\\.approved"), "");
    }
}
